asm_to_pil/src/lib.rs

```diff
@@ -1,1081 +1,129 @@
-        FunctionDefinition, PilStatement, PolynomialName, PolynomialReference, SelectedExpressions,
-        UnaryOperator,
-    },
-};
-
-const MAIN: &str = "Main";
-
+use ast::asm_analysis::AnalysisASMFile;
+// use function_desugar::desugar_machine;
 use number::FieldElement;
+use romgen::generate_machine_rom;
 
-pub fn compile<T: FieldElement>(analysis: AnalysisASMFile<T>) -> PILGraph<T> {
-    Session::default().compile(analysis)
-}
-
-#[derive(Default)]
-/// A compilation session
-struct Session<T> {
-    /// the machine types used in the tree
-    machine_types: BTreeMap<String, Machine<T>>,
-}
-
-impl<T: FieldElement> Session<T> {
-    fn instantiate_machine(&mut self, location: &Location, ty: String) -> Object<T> {
-        ASMPILConverter::new(location, &self.machine_types)
-            .convert_machine(self.machine_types.get(&ty).unwrap().clone())
-    }
-
-    fn compile(&mut self, mut input: AnalysisASMFile<T>) -> PILGraph<T> {
-        let main_location = Location::default().join("main");
-
-        // we start from the main machine
-        let main_ty = match input.machines.len() {
-            // if there is a single machine, treat it as main
-            1 => input.machines.keys().next().unwrap().clone(),
-            // otherwise, use the machine called `MAIN`
-            _ => {
-                assert!(input.machines.contains_key(MAIN));
-                MAIN.into()
-            }
-        };
-
-        self.machine_types = std::mem::take(&mut input.machines);
-
-        // get a list of all machines to instantiate. The order does not matter.
-        let mut queue = vec![(main_location.clone(), main_ty)];
-
-        let mut instances = vec![];
-
-        while let Some((location, ty)) = queue.pop() {
-            let machine = self.machine_types.get(&ty).unwrap();
-
-            queue.extend(machine.submachines.iter().map(|def| {
-                (
-                    // get the absolute name for this submachine
-                    location.clone().join(def.name.clone()),
-                    // get its type
-                    def.ty.clone(),
-                )
-            }));
-
-            instances.push((location, ty));
-        }
+mod common;
+mod romgen;
+mod vm_to_constrained;
 
-        // visit the tree compiling the machines
-        let objects = instances
+/// Remove all ASM from the machine tree. Takes a tree of virtual of constrained machines and returns a tree of constrained machines
+pub fn compile<T: FieldElement>(file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
+    AnalysisASMFile {
+        machines: file
+            .machines
             .into_iter()
-            .map(|(location, ty)| {
-                let object = self.instantiate_machine(&location, ty);
-                (location, object)
+            .map(|(name, m)| {
+                (name, {
+                    let (m, rom) = generate_machine_rom(m);
```

Comment:
I don't agree. I'm trying to rely as little as possible on runtime assumptions. I agree it's a bit awkward to have a single type for both VMs and CMs but prefer relying on the type of `generate_machine_rom` (which returns an optional ROM) rather than a runtime check inside its implementation.